crates/bevy_ecs/src/query/state.rs

```diff
@@ -778,6 +811,157 @@ impl<Q: WorldQuery, F: ReadOnlyWorldQuery> QueryState<Q, F> {
             );
         }
     }
+    /// This is a batched version of `for_each_mut` that accepts a batch size, `N`, and a desired alignment for the batches `ALIGN`.
+    /// Runs `f` and `f_batch` on each query result for the given [`World`].  This can be used to improve performance of your queries using SIMD.
+    ///
+    /// Usage:
+    ///
+    /// * Supported values of `ALIGN` are 16, 32, and 64
+    /// * `N` must be a power of 2
+    /// * `f_batch` receives components in "batches" that are aligned to `ALIGN`.
+    /// * `f` functions exactly as does in [`QueryState<Q,F>::for_each_mut`] -- it receives "scalar" (non-batched) components.
+    ///
+    /// In other words, `f_batch` composes the "fast path" of your query, and `f` is the "slow path".
+    ///
+    /// Batches are currently one of [`bevy_ptr::batch::AlignedBatch16`], [`bevy_ptr::batch::AlignedBatch32`], [`bevy_ptr::batch::AlignedBatch64`] types, each corresponding to a guaranteed batch alignment.
+    /// The batch alignment is important as it enables architecture-specific optimizations that depend on alignment.
+    ///
+    /// If you attempt to call this function with an invalid generic argument, a compile-time error will be issued.  For example,
+    /// if you request 64-bytes alignment on a batch whose size is 8 bytes, your program will fail to compile.  In general,
+    /// choose `ALIGN` as large as you can for your batch parameters.  The minimum alignment supported for batches is 16 bytes.
+    ///
+    /// **Note**: Once generalized const expressions are stable, it will be possible to compute `ALIGN` automatically based on the component type and batch size.
+    /// Compile-time assertions will still exist for sanity checking.
+    ///  
+    /// In general, when using this function, be mindful of the types of filters being used with your query, as these can fragment your batches
```

Comment:
Interesting advice, but needs to be a bit more concrete to be actionable.

Reply:
I'm thinking about putting an example here with change trackers -- the query should still be "dense" but possibly highly fragmented.  The effect of that is that the "scalar path" has to run a lot more often due to the incomplete batches available.

For example (forgive the ASCII art):

~~~
batch =   | b b b b b b b b |
filter => | y y y y n y y y | 

~~~

In this example, a would-be batch is "fragmented" because the filter yielded `false` for one element of the batch.  Therefore, the scalar path has to be executed for all elements of the batch, which is no better than `for_each_mut`.

But a more friendly query filter might be something like `With<Component>`, which basically just restricts which tables are included in the query.  That is, the query itself isn't "fragmented" anymore.  Elements won't just spuriously return `false` in the middle of iteration of a table -- the filter will either succeed for the entire table, or fail for the entire table.

I believe these are considered "Archetypal" queries in Bevy but I haven't found much documentation on that.

Btw, even with "friendlier" filters, the initial batching implementation isn't that smart yet... I expect performance to not be super great with them at the moment.  However, the interface is at least built such that code using that would get faster as better implementations are made.  For now I think the "avoid filters" advice is the most practical, until we get those smarts into the batch iteration code anyway.

Reply:
Good, I think this is a great start.